File: service_backend/app/schemas/integration.py
"""Integration schemas (plan 09 §6) — camelCase on the wire.

Credentials are WRITE-ONLY: accepted on create/update, never serialized back.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import Field

from app.schemas.base import ApiModel


class ProviderOut(ApiModel):
    """Catalog entry — `fields` is the config schema driving the wizard."""

    provider: str
    type: str
    title: str
    description: str
    icon: Optional[str] = None
    fields: List[Dict[str, Any]]
    testLabel: str
    testTarget: Optional[Dict[str, str]] = None


class ConnectionOut(ApiModel):
    id: str
    tenantId: str
    provider: str
    type: str
    name: str
    # Non-secret config only — camelCase keys for the frontend.
    config: Dict[str, str]
    status: str
    lastTestedAt: Optional[datetime] = None
    lastError: Optional[str] = None
    rateLimitPerMinute: int
    createdAt: datetime
    updatedAt: datetime


class ConnectionCreateRequest(ApiModel):
    provider: str = Field(min_length=1, max_length=60)
    name: str = Field(min_length=1, max_length=120)
    config: Dict[str, str] = Field(default_factory=dict)
    # Write-only secrets; on update an omitted key keeps the stored value.
    credentials: Dict[str, str] = Field(default_factory=dict)
    rateLimitPerMinute: int = Field(default=30, ge=1, le=600)


class ConnectionUpdateRequest(ApiModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=120)
    config: Optional[Dict[str, str]] = None
    credentials: Optional[Dict[str, str]] = None
    rateLimitPerMinute: Optional[int] = Field(default=None, ge=1, le=600)


class ConnectionListResponse(ApiModel):
    """Resource-shell page (plan 06 D6) — like every list entity."""

    data: List[ConnectionOut]
    total: int
    page: int


class ConnectionNeighborResponse(ApiModel):
    """Record-nav: the connection at `index` under the carried query."""

    connection: Optional[ConnectionOut] = None
    total: int


class TestRequest(ApiModel):
    # Optional targeted test (SMTP: recipient). Absent = connection check.
    target: Optional[str] = Field(default=None, max_length=254)


class TestResultOut(ApiModel):
    ok: bool
    message: str
    checkedAt: datetime

File: service_frontend/types/integration.ts
/**
 * Integration framework types (plan 09) — the core "connections" primitives
 * every external-service integration plugs into. SMTP (email) is the first
 * provider; future providers (storage, LLM, ERP) reuse these shapes, whether
 * they ship in core or as App Store modules.
 */

/** Integration category — drives grouping/iconography, not behavior. */
export type IntegrationType =
  | 'email'
  | 'storage'
  | 'llm'
  | 'erp'
  | 'payment'
  | 'consumer'
  // Meetings module (S0): the calendar reader and the notetaker bot account.
  // Two categories, not one, because the one-active-per-type invariant must
  // let a tenant hold BOTH at once.
  | 'calendar'
  | 'meeting_bot';

/** Connection health — UNVERIFIED until a test passes, ERROR on failures. */
export type ConnectionStatus = 'ACTIVE' | 'UNVERIFIED' | 'ERROR';

/** One wizard form field, declared by the provider (drives the configure step). */
export interface ProviderField {
  /** Key in `config` (or `credentials` when `secret`). */
  key: string;
  label: string;
  type: 'text' | 'number' | 'password' | 'select';
  required: boolean;
  /** Secret values go to encrypted `credentials` and are write-only. */
  secret?: boolean;
  placeholder?: string;
  /** Select options (type === 'select'). */
  options?: Array<{ value: string; label: string }>;
  defaultValue?: string;
  /** Collapsed under the wizard's "Advanced" section. */
  advanced?: boolean;
}

/** Catalog entry (GET /integrations/providers) — config schema drives the wizard. */
export interface IntegrationProvider {
  /** API identity, e.g. "smtp". */
  provider: string;
  type: IntegrationType;
  title: string;
  description: string;
  /** Lucide icon key (graceful fallback for unknowns). */
  icon: string | null;
  fields: ProviderField[];
  /** Label for the OPTIONAL targeted test, e.g. "Send test email". The default
   *  test is always a plain connection check (connect + authenticate). */
  testLabel: string;
  /** Present = the provider offers an optional targeted test (SMTP: send a
   *  real email to an address). Null = connection check only. */
  testTarget: { label: string; placeholder: string } | null;
}

/** A tenant's configured connection — credentials are NEVER returned. */
export interface Connection {
  id: string;
  tenantId: string;
  provider: string;
  type: IntegrationType;
  name: string;
  /** Non-secret config only (host, port, fromEmail, …). */
  config: Record<string, string>;
  status: ConnectionStatus;
  /** Storage write-target flag — exactly one active storage connection per
   * tenant; new uploads land there, resolve-by-type serves it. */
  isActive: boolean;
  lastTestedAt: string | null; // ISO
  lastError: string | null;
  /** Outbox dispatcher throttle (plan 09 §5 — low-spec SMTP guard). */
  rateLimitPerMinute: number;
  createdAt: string; // ISO
  updatedAt: string; // ISO
}

/** Create/update payload — secrets travel separately, write-only. */
export interface ConnectionInput {
  provider: string;
  name: string;
  config: Record<string, string>;
  /** Encrypted at rest; on update an OMITTED key means "keep existing". */
  credentials: Record<string, string>;
  rateLimitPerMinute?: number;
}

/** Inline test outcome (wizard test step / card Test action). */
export interface TestConnectionResult {
  ok: boolean;
  message: string;
  checkedAt: string; // ISO
}

/** Display state for the integration card StatusBadge (derived, not stored). */
export type ConnectionBadge = 'NOT_CONNECTED' | 'CONNECTED' | 'UNVERIFIED' | 'ERROR';

export function connectionBadge(c: Connection | undefined): ConnectionBadge {
  if (!c) return 'NOT_CONNECTED';
  if (c.status === 'ACTIVE') return 'CONNECTED';
  return c.status;
}
